### PR TITLE
Multi use docker

### PR DESCRIPTION
**Pull request by rhliang**
Convert the `micall_basespace.py` driver to a more all-purpose driver that enables the user to process a full run *or* a single sample.

This driver (now called `micall_docker.py`) has four modes of operation:

* `./micall_docker.py basespace` processes a BaseSpace run;
* `./micall_docker.py folder` will process a MiSeq run outside of a BaseSpace setting;
* `./micall_docker.py sample` processes a single sample, which you specify via command-line parameters;
* `./micall_docker.py hcv_sample` processes a single HCV sample, also specified via command-line parameters.

### CODE REVIEW

**Comment by donkirkby** [comment]
Suggestions:

* Rename `micall_basespace.py` to `micall_docker.py`. (I think you suggested that the other day, and it's a good idea.)
* Add another subcommand: `basespace`. Since it's better to be explicit than implicit, let's remove the guesswork where it looks for the basespace file and then falls back to the run folder. I think that means that we don't need the `--link_run` option. `basespace` will always read from `/input`, and `folder` can have a positional argument that defaults to `/input`.
* Move the help text from the Dockerfile to the argparse help text. In the Dockerfile, just put a comment that says `docker run micall:version --help` for instructions.

File: micall_docker.py [hunk 1]
```diff
@@ -985,9 +982,8 @@ def link_samples(
         run_path: str,
         data_path: str,
         is_denovo: bool,
-        fastq1s: Optional[Iterable[str]],
-        fastq2s: Optional[Iterable[str]],
-):
+        fastq1s: typing.Sequence[str] = None,
```
Comment (nathanielknight) [inline micall_docker.py:985]: Should this not be `Optional[Sequence[str]]`?

Comment (donkirkby) [reply micall_docker.py:985]: I thought the `= None` implied the `Optional`. PyCharm wasn't complaining, but I haven't checked the rules or tried mypy.

Comment (nathanielknight) [reply micall_docker.py:985]: Huh, so it does. Today I Learned ([docs, for future reference](https://docs.python.org/3/library/typing.html#typing.get_type_hints))

**Pull request merged**